[pigeon] moved to a custom codec

gaaclarke opened the pull request:
Now a custom MessageCodec will be generated for each declared API in the pigeon file instead of using the StandardMessageCodec.  This ensures that the messages are recursively encoded and that custom classes inside of collections like Maps or List will serialize correctly.  This will also open the door to make primitive datatypes easier to implement.

fixes https://github.com/flutter/flutter/issues/80538

## Pre-launch Checklist

- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [ ] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated `CHANGELOG.md` to add a description of the change.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the `#hackers-new` channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning

gaaclarke commented:
@stuartmorgan I have a problem where the analyze bot dies on this PR because it is analyzing code that depends on generated code that is no longer checked in.  It seems like we could add a step before analysis happens to generate code, or we exclude that directory from analysis somehow and just rely on local_tests to perform the analysis after the code is generated.  What do you think?

stuartmorgan commented:
> @stuartmorgan I have a problem where the analyze bot dies on this PR because it is analyzing code that depends on generated code that is no longer checked in.

Can we discuss this requirement further? Why does Pigeon itself need generated code in order to run?

> It seems like we could add a step before analysis happens to generate code

If this ends up being a hard requirement, it would be simple to make a new tools command that runs an arbitrary file if present, and then add a step to CI to run it.

That could also replaces the existing "local test" shell script.

> or we exclude that directory from analysis somehow and just rely on local_tests to perform the analysis after the code is generated.

This would move Pigeon toward being even more bespoke and inconsistent, which add I've said is the opposite of how we should be doing CI.

stuartmorgan commented on packages/pigeon/lib/dart_generator.dart:
```diff
@@ -49,8 +49,58 @@ String _escapeForDartSingleQuotedString(String raw) {
       .replaceAll(r"'", r"\'");
 }
 
+String _calcCodecName(Api api) => '_${api.name}Codec';
```
What is `calc` trying to indicate here that is different from `get`? This seems like an odd name.

gaaclarke replied:
I usually only use `get` when it's idiomatic for the language or when I'm retrieving a cached value.  To me `calc` means a pure function that will actually calculate something rather than look it up.  I think some of that messiness is also why some languages did away with the "get" verb for getters.

stuartmorgan replied:
I'd encourage you to reconsider this philosophy; I have never encountered that convention, so I have no idea what `calc` means when I read it, and I doubt very much that I'm alone in that. Using a name that means something special to you, but is somewhere between odd and actively confusing to a reader who isn't you doesn't improve readability overall. I found this actively harder to read than if it had been called `_getCodecName`, since I had to stop to figure out what `calc` was supposed to mean (and couldn't).

I'm all for using better verbs than "get" when there's a clear, more precise alternative, but avoiding "get" just to avoid it can make the code less clear.

gaaclarke replied:
"Calc" is listed as an abbreviation for "calculate" at dictionary.com: https://www.dictionary.com/browse/calc

Looks like google's style guide actually sets the bar for an abbreviation if it is listed on wikipedia.  Wikipedia says for "calc" "Short for calculation, calculator, or calculus".

It doesn't seem that confusing to me and looks like a grey area in the style guide.  If it threw you off I imagine it could be even more confusing for a non-native english speaker.

gaaclarke replied:
done

stuartmorgan replied:
> "Calc" is listed as an abbreviation for "calculate" at dictionary.com: https://www.dictionary.com/browse/calc
> 
> Looks like google's style guide actually sets the bar for an abbreviation if it is listed on wikipedia. Wikipedia says for "calc" "Short for calculation, calculator, or calculus".

It sounds like you misunderstood my concern. I understood that "calc" was short for "calculate". What I didn't understand was why it was called "calculate" rather than "get".

If someone uses an unusual verb—and this is unusual, because I have never heard someone say that they were going to "calculate" a string, especially for trivial concatenation—then it makes me assume they are trying to communicate that something unusual is happening in that method.

stuartmorgan commented on packages/pigeon/lib/dart_generator.dart:
```diff
@@ -49,8 +49,58 @@ String _escapeForDartSingleQuotedString(String raw) {
+        bool first = true;
+        for (final EnumeratedClass customClass in getCodecClasses(api)) {
+          indent.write(
+              '${first ? '' : 'else '}if (value is ${customClass.name}) ');
```
You could just output 'else ' at the end of the loop, instead of conditionally here and also just after the loop.

gaaclarke replied:
The reason it is here is because we only want an else clause if `getCodecClasses(api).length > 0`.  This allows us to avoid invoking `getCodecClasses` twice.  Pulling out the else clause until after the loop would still require some variable to check that condition or calling the function twice.  It wouldn't be any cleaner imo.

stuartmorgan replied:
I don't follow. Your current code is:
```
        for (final EnumeratedClass customClass in getCodecClasses(api)) {
          indent.write(
              '${first ? '' : 'else '}if (value is ${customClass.name}) ');
          [...]
        }
        indent.write('else ');
```

Doing something at the beginning of every loop on every iteration after the first and also unconditionally after the loop identical to just doing something at the end of the loop. What extra variable would be required?

gaaclarke replied:
Ahh, I remember what happened, sorry I was misremembering this.

Originally the code was written such that the `first` variable was used to decide if an else clause was needed.  Then I went back and just removed the codegen for the whole method if `getCodecClasses(api).length <= 0`.  After that it was logically correct so I didn't go back and shuffle it around.

The fact that you brought it up a few times leads me to believe you find it confusing, I'll refactor it.

gaaclarke replied:
done

stuartmorgan commented on packages/pigeon/lib/generator_tools.dart:
```diff
@@ -249,3 +249,42 @@ Map<String, Object> mergeMaps(
+  final int enumeration;
+}
+
+/// Custom codec's custom types are enumerated from 127 down to this number to
```
Typo: `codecs'`

gaaclarke replied:
done

stuartmorgan commented on packages/pigeon/lib/generator_tools.dart:
```diff
@@ -249,3 +249,42 @@ Map<String, Object> mergeMaps(
+  final int enumeration;
+}
+
+/// Custom codec's custom types are enumerated from 127 down to this number to
```
Why down?

gaaclarke replied:
It's more resistant to future changes in the StandardMessageCodec.  If we add a field to the StandardMessageCodec and someone starts using an old version of Pigeon with it, a runtime error is less likely: P(add field to StandardMessageCodec) * P(user has more than 117 referenced datatypes in an API).

gaaclarke replied:
This changed now that I'm in the 128-255 range.  Now it's pretty arbitrary that I'm guaranteed not to collide.

stuartmorgan commented on packages/pigeon/lib/generator_tools.dart:
```diff
@@ -249,3 +249,42 @@ Map<String, Object> mergeMaps(
+  sortedNames.sort();
+  int enumeration = 127;
+  for (final String name in sortedNames) {
+    if (enumeration < _minimumCodecFieldKey) {
```
You know the number of entries before starting the loop; why not just check once outside the loop that there aren't too many?

gaaclarke replied:
done

stuartmorgan commented on packages/pigeon/lib/java_generator.dart:
```diff
@@ -64,6 +64,59 @@ class JavaOptions {
+        bool first = true;
+        for (final EnumeratedClass customClass in getCodecClasses(api)) {
+          indent.write(
+              '${first ? '' : 'else '}if (value instanceof ${customClass.name}) ');
```
Same comment here about 'else '

gaaclarke replied:
addressed above

stuartmorgan commented on packages/pigeon/lib/objc_generator.dart:
```diff
@@ -134,6 +134,88 @@ void _writeClassDeclarations(
+      bool first = true;
+      for (final EnumeratedClass customClass in getCodecClasses(api)) {
+        indent.write(
+            '${first ? '' : 'else '}if ([value isKindOfClass:[${_className(options.prefix, customClass.name)} class]]) ');
```
And here.

gaaclarke replied:
addressed above

stuartmorgan commented on packages/pigeon/lib/generator_tools.dart:
```diff
@@ -249,3 +249,42 @@ Map<String, Object> mergeMaps(
+}
+
+/// Custom codec's custom types are enumerated from 127 down to this number to
+/// avoid collisions with the StandardMessageCodec.
```
This isn't avoid collisions, it's violating the documented requirements such that it will *create* collisions. You shouldn't be using anything in this range.

> Type discriminators 0 through 127 inclusive are reserved for use by the base class, as follows:
> - [...]
> - Reserved for future expansion: 14..127
>
> The codec can be extended by overriding this method, calling super for values that the extension does not handle. Type discriminators used by extensions must be greater than or equal to 128 in order to avoid clashes with any later extensions to the base class.

gaaclarke replied:
Yep, I read this wrong.  I'll change it, good catch.

gaaclarke replied:
done

stuartmorgan commented on packages/pigeon/platform_tests/ios_unit_tests/ios/RunnerTests/AsyncHandlersTest.m:
```diff
@@ -8,24 +8,25 @@
-+ (Value*)fromMap:(NSDictionary*)dict;
-- (NSDictionary*)toMap;
++ (Value *)fromMap:(NSDictionary *)dict;
+- (NSDictionary *)toMap;
```
Is there a reason you're changing the style of this file?

gaaclarke replied:
I just do whatever the CI tells me for this.  Maybe the clang-tidy was upgraded on the bot?

gaaclarke replied:
Lookingat this again, this looks unnecessary.  Let me try to remove it from the patch.

gaaclarke replied:
removed

gaaclarke commented:
@stuartmorgan I think I addressed your concerns.  The most important thing was switching the range that the type identifiers is in.  The other comments were addressed except for a few where I offered explanations.  Let me know if there is anything else.

stuartmorgan commented on packages/pigeon/lib/generator_tools.dart:
```diff
@@ -249,3 +249,42 @@ Map<String, Object> mergeMaps(
+  final int enumeration;
+}
+
+/// Custom codecs' custom types are enumerated from 255 down to this number to
```
Iterating down rather than up seems odd now that you're no longer in a contested ID space.

gaaclarke replied:
done

stuartmorgan commented on packages/pigeon/lib/objc_generator.dart:
```diff
@@ -134,6 +134,85 @@ void _writeClassDeclarations(
+String _getCodecName(String? prefix, String className) =>
+    '${_className(prefix, className)}Codec';
+
+String _calcCodecGetterName(String? prefix, String className) =>
```
This should be changed as well, as the inconsistency with `_getCodeName` exacerbates my readability concern: the reader is now left to try to understand (in vain, because there's no reason) why two methods that have identical construction use two different verbs.

gaaclarke replied:
done

stuartmorgan reviewed (approved):
LGTM with nits.
